dotnet-test-samples/kinesis-lambda-dynamodb/tests/KinesisEventHandler.UnitTests/Handlers/KinesisEventHandlerTests.cs

```diff
@@ -0,0 +1,90 @@
+using System.Threading.Tasks;
+using Amazon.Lambda.Core;
+using Amazon.Lambda.TestUtilities;
+using FluentAssertions;
+using KinesisEventHandler.Handlers;
+using KinesisEventHandler.Models;
+using KinesisEventHandler.UnitTests.Utilities;
+using Moq;
+using Xunit;
+
+namespace KinesisEventHandler.UnitTests.Handlers;
+
+public class KinesisEventHandlerTests
+{
+    private readonly Mock<KinesisEventHandler<Employee>> _mockKinesisEventTrigger;
+
+    public KinesisEventHandlerTests()
+    {
+        _mockKinesisEventTrigger = new Mock<KinesisEventHandler<Employee>>();
+
+        _mockKinesisEventTrigger.Setup(x =>
+                x.ProcessKinesisRecord(It.IsAny<Employee>(), It.IsAny<ILambdaContext>()))
+            .Returns(Task.CompletedTask);
+    }
+
+    [Fact]
+    public async Task KinesisEventHandler_Should_CallProcessKinesisRecordOnce()
+    {
+        //Arrange
+        var expected = new EmployeeBuilder().Build();
+        var kinesisEvent = new KinesisEventBuilder().WithEmployees(new[] { expected });
+        var lambdaContext = new TestLambdaContext();
+
+        //Act
+        var result = await _mockKinesisEventTrigger.Object.Handler(kinesisEvent, lambdaContext);
```

Comment:
The object under test and mock are the same this is an anti-pattern, how come we're faking the same object we're testing?
It means we should fake one of the unit's dependencies or refactor the code

Reply:
KinesisEventHandler is an abstract class that the Function class implements, we're mocking the implementation of the abstract method and expecting a certain behavior i.e. the implemented method is called only certain number of times based on count of incoming records in Kinesis event. It's not an anti-pattern when done within the context of the handler. Thoughts?

Reply:
I would still argue that we don't want to test that a function was called, it would bind the test to a specific implementation, instead it's preferable where possible to test a result - in other words that the data was stored/sent

Reply:
I think we need to test the calls because we're testing a pattern that supports batch processing. Kinesis event can contain `n` number of records and the way we have structured the abstraction is that methods are meant to be called for each record within the Kinesis event. Future developers can override this can manually call the methods, the tests here will ensure that methods are called correctly and someone is not overriding this behavior.